### DISCUSSION TITLE
Enable git clean/smudge filters when checking out a git repository

### DISCUSSION THREAD

**Discussion by mikaello** [renovatebot/renovate#17932] [Suggest an Idea]
**UPDATE, TLDR:** You can provide your own Git config by creating a custom image off Renovate's image and adding your config to `/home/ubuntu/.gitconfig`. For smudge filters that affects dependencies this is not sufficient/will not work since Renovate performs additional direct API calls to your Git instance to fetch package files, and these calls will not use the smudge filters.

### What would you like Renovate to be able to do?

In our setup we need to use Git clean and smudge filters for our repositories, because the dependencies are fetched from different remotes depending on who's checking out the code and which firewall their behind. This makes it necessary to alter the source code with a smudge filter before it makes sense to Renovate, and it needs to be cleaned before pushing back to make sense to other participants later on.

This is achievable with[ Git "smudge" and "clean" filter](https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes#filters_a) (see also [filter documentation for gitattributes](https://git-scm.com/docs/gitattributes#_filter)), see also this article for an explanation: [Protect secrets in Git with the clean/smudge filter
](https://developers.redhat.com/articles/2022/02/02/protect-secrets-git-cleansmudge-filter).

This is easily done locally by adding the config like this:

```shell
git config filter.myFilterName.clean "sed -e 's/text in Git remote/text locally/g'"
git config filter.myFilterName.smudge "sed -e 's/text locally/text in Git remote/g'"
```

but not so in Renovate, because the config is added inside Renovate with the `simple-git` dependency.

### If you have any ideas on how this should be implemented, please tell us here.

I am not quite sure what is the best way to add this, as the config for each filter may consist of three parts (depending on how flexible this should be): filter name ("driver"), filter type (`smudge` or `clean`) and the filter itself. A possible implementation of the actual setting of the filter could be (modelled after [`writeGitAuthor`](https://github.com/renovatebot/renovate/blob/54b36695a7b965e2a61a6141c74e71a2ca3fb7dd/lib/util/git/index.ts#L283-L310)):

```typescript
export async function writeGitFilters(
  filters: {
    filterName: string, // Git terminology is "driver"
    filterType: 'smudge' | 'clean',
    filterContent: string,
  }[] = []
): Promise<void> {
  // istanbul ignore if
  if (config.writeGitFilterDone) {
    return;
  }
  config.writeGitFilterDone = true;

  for (const filter of filters) {
    try {
      logger.debug(filter, "Setting git filter");
      await git.addConfig(
        `filter.${filter.filterName}.${filter.filterType}`,
        filter.filterContent
      );
    } catch (err) /* istanbul ignore next */ {
      const errChecked = checkForPlatformFailure(err);
      if (errChecked) {
        throw errChecked;
      }
      logger.debug({ err, ...filter }, "Error setting git filter config");
      throw new Error(TEMPORARY_ERROR);
    }
  }
}
```

This seems difficult to achieve by environment variables.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
Could it be represented with new config options like `gitSmudges` and `gitFilters` which are simple string->string objects mapping before to after?

**Comment by mikaello** [comment]
> Could it be represented with new config options like gitSmudges and gitFilters which are simple string->string objects mapping before to after?

Hm, yes, that could work. Something like:

```json
{
  "gitSmudgeFilters": {
    "myFilter": "sed -e 's/text in Git remote/text locally/g'",
    "catSmudgeFilter": "cat"
  },
  "gitCleanFilters": {
    "myFilter": "sed -e 's/text locally/text in Git remote/g'",
    "indentCleanFilter": "indent"
  }
}
```

_Notice:_ in Git it is also possible to define a third type, [a long running filter process](https://git-scm.com/docs/gitattributes#_long_running_filter_process): `filter.<driver>.process` (in comparison to `filter.<driver>.smudge` and `filter.<driver>.clean`). This is just for information, and it should be easy to extend the config above with a new property to support that if that is desirable in the future.

**Comment by viceice** [comment]
those needs to be global only options for self-hosted renovate, as it can execute arbitrary code.

**Comment by mikaello** [comment]
I have been thinking a bit, and I have a proposal for an alternative config structure:

## Proposal

```json
{
  "gitFilterDrivers": [
    {
      "name": "myFilter",
      "actions": [
        {
          "type": "smudge",
          "value": "sed -e 's/text in Git remote/text locally/g"
        },
        {
          "type": "clean",
          "value": "sed -e 's/text locally/text in Git remote/g"
        },
        {
          "type": "required",
          "value": "true"
        }
      ]
    },
    {
      "name": "catSmudgeFilter",
      "actions": [{ "type": "smudge", "value": "cat" }]
    },
    {
      "name": "indentCleanFilter",
      "actions": [{ "type": "clean", "value": "indent" }]
    }
  ]
}
```

This is more verbose, but will resemble the Git config more closely, and it will also support the type `required` and `process` (and any other additions to filter that may come). `required` is documented as such:

> If a filter must succeed in order to make the stored contents usable, you can declare that the filter is `required`

Source: https://git-scm.com/docs/gitattributes#_filter

I realise that this may be more complicated, and maybe you have some convention about not having more nested structures than necessary. Just wanted to hear your take on this, before an attempt at implementing is done.

## Rationale

When running the following commands:

```shell
$ git config filter.myFilter.smudge "sed -e 's/text in Git remote/text locally/g'"
$ git config filter.catSmudgeFilter.smudge cat
$ git config filter.myFilter.clean "sed -e 's/text locally/text in Git remote/g'"
$ git config filter.indentCleanFilter.clean indent
$ git config filter.myFilter.required true
```

... the following entries are created in my `.git/config` file:

```gitconfig
[filter "myFilter"]
        smudge = sed -e 's/text in Git remote/text locally/g'
        clean = sed -e 's/text locally/text in Git remote/g'
        required = true
[filter "catSmudgeFilter"]
        smudge = cat
[filter "indentCleanFilter"]
        clean = indent
```

As you can see, all actions / properties (`smudge`, `clean`, `required`) are grouped under a filter driver definition (filter name).

Following the same structure in the config will increase familiarity and make it more obvious what actions/properties are belonging to a certain filter driver definition.

**Comment by rarkins** [comment]
If you are running self-hosted, does anything stop you simply creating the `~/.git/config` file?

**Comment by mikaello** [comment]
Before I created this issue, I did quite a lot of testing (copy our own Git config into our custom Renovate container), but with no success, and I settled with that thought that `simple-git` did not read user config.

But before answering you now, I did some more testing, and it seems I used the wrong path, because now I got Renovate to obey our Git config. The correct way of adding custom Git config is to place the config at `/home/ubuntu/.gitconfig`:

```Dockerfile
COPY config /home/ubuntu/.gitconfig
```

This works half the way, but from the new error message, it seems like Renovate are doing some fetching of resources without using Git clone? Because now it is failing with: 

```json
{
  "level": 20,  "repository": "myRepo",  "branch": "renovate-myDep",
  "packageFile": "kustomization/kustomization.yaml",
  "depName": "artifactory.company.com/github-docker-remote/myDep",
  "msg": "Found match at index 273",
}
{
  "level": 20,  "repository": "myRepo",  "branch": "renovate-myDep",
  "manager": "kustomize",
  "packageFile": "kustomization/kustomization.yaml",
  "currentDepName": "artifactory.company.com/github-docker-remote/myDep",
  "newDepName": "artifactory.company.com/github-docker-remote/GIT_NAME",
  "msg": "depName mismatch",
}
{
  "level": 40,  "repository": "myRepo",  "branch": "renovate-myDep",
  "msg": "Error updating branch: update failure",
}
```

Where `myDep` is what it should be when checked out, and `GIT_NAME` is what is stored in Git, i.e.

```gitconfig
[filter "myFilter"]
        smudge = sed -e 's|GIT_NAME|myDep|g'
        clean = sed -e 's|myDep|GIT_NAME|g'
        required = true
```


Location of error message:

https://github.com/renovatebot/renovate/blob/4168b34e7b37a195b406c93984d94069ca60cf77/lib/workers/repository/update/branch/auto-replace.ts#L47-L57

**Reply by rarkins** [reply]
Is this perhaps from datasource querying?

**Reply by mikaello** [reply]
Hm, I am sorry, now I don't follow you.

From what I can see, Renovate is obeying the filters at checkout, since it actually converted `artifactory.company.com/github-docker-remote/GIT_NAME` to `artifactory.company.com/github-docker-remote/myDep`, but somehow it finds the name `artifactory.company.com/github-docker-remote/GIT_NAME` (which is an internal name only usable in an local/internal context). This internal name is not present in the datasource, and the only source for this name is by looking into the source code in GitLab (without applying smudge filters, so a `GET` request using the GitLab API fetching the file directly would see this internal name since it will not use smudge filter).

**Reply by rarkins** [reply]
There are definition scenarios such as the [`gitlab-tags`](https://github.com/renovatebot/renovate/tree/main/lib/modules/datasource/gitlab-tags) datasource where Renovate uses the GitLab API. You should search your debug logs for:
- All instances of the correct/smudged name
- First instance of the incorrect/raw name

**Reply by mikaello** [reply]
Thanks a lot for your patience.

* All instances of correct/smudge name seems normal, Renovate finds it as an update in `config.kustomize[].deps[].depName` in log message `packageFiles with updates`, checks for existing PR before the log messages above that results in `depName mismatch`
* First and only instance of incorrect/raw name is the log message [pasted above](https://github.com/renovatebot/renovate/discussions/17932#discussioncomment-3714202)

I have now tried with two other managers, a regex-manager and a npm-manager, both of these fails, but with slightly different but related errors, and in both cases they got hold of the content from Git without it being smudged:

<details><summary><strong>REGEX:</strong> First and only log message mentioning incorrect/raw name (<code>GGG_RAW</code>, correct name is <code>gantsign</code>)</summary><p>

Field `existingContent` contains un-smudged/raw text, while `replaceString` has correct name:

```json
{
  "name": "renovate",  "level": 30,  "repository": "myRepo",  "branch": "renovate-com.github.gantsign.maven-ktlint-maven-plugin-1.x",
  "packageFile": "bash/file.sh",
  "depName": "com.github.gantsign.maven:ktlint-maven-plugin",
  "existingContent": "# renovate: datasource=maven depName=com.github.GGG_RAW.maven:ktlint-maven-plugin\nexport KTLINT_MAVEN_PLUGIN_VERSION=1.14.0\nalias ktlintf='mvn com.github.GGG_RAW.maven:ktlint-maven-plugin:${KTLINT_MAVEN_PLUGIN_VERSION}:format'",
  "replaceString": "renovate: datasource=maven depName=com.github.gantsign.maven:ktlint-maven-plugin\nexport KTLINT_MAVEN_PLUGIN_VERSION=1.14.0\n",
  "msg": "Cannot find replaceString in current file content. Was it already updated?"
}
```

</p></details>


<details><summary><strong>NPM:</strong> First and only log message mentioning incorrect/raw name (<code>DDD_RAW</code>, correct name is <code>@datapunt/matomo-tracker-react</code>)</summary><p>

As you can see, `existingContent` has raw content from Git (containing `DDD_RAW`), while rest of the log message uses correct name `@datapunt/matomo-tracker-react`:

```json
{
  "branch": "renovate-datapunt-matomo-tracker-react-0.x",
  "level": 20,
  "repository": "myRepo",
  "msg": "Error updating file",
  "existingContent": {
    "name": "npm-test",
    "version": "1.0.0",
    "main": "index.js",
    "type": "module",
    "dependencies": {
      "DDD_RAW": "0.3.0"
    }
  },
  "config": {
    "versioning": "npm",
    "branchName": "renovate-datapunt-matomo-tracker-react-0.x",
    "commitMessage": "Update dependency @datapunt/matomo-tracker-react to v0.5.1",
    "prTitle": "Update dependency @datapunt/matomo-tracker-react to v0.5.1",
    "language": "js",
    "manager": "npm",
    "packageFile": "npm/package.json",
    "packageJsonName": "npm-test",
    "packageFileVersion": "1.0.0",
    "packageJsonType": "library",
    "npmLock": "npm/package-lock.json",
    "lockFiles": ["npm/package-lock.json"],
    "parentDir": "npm",
    "packageFileDir": "npm",
    "depType": "dependencies",
    "depName": "@datapunt/matomo-tracker-react",
    "currentValue": "0.3.0",
    "datasource": "npm",
    "prettyDepType": "dependency",
    "lockedVersion": "0.3.0",
    "sourceUrl": "https://github.com/Amsterdam/matomo-tracker",
    "currentVersion": "0.3.0",
    "fixedVersion": "0.3.0",
    "bucket": "non-major",
    "newVersion": "0.5.1",
    "newValue": "0.5.1",
    "depNameSanitized": "datapunt-matomo-tracker-react",
    "sourceRepoSlug": "Amsterdam-matomo-tracker",
    "sourceRepo": "Amsterdam/matomo-tracker",
    "sourceRepoOrg": "Amsterdam",
    "sourceRepoName": "matomo-tracker",
    "baseDeps": [
      {
        "depType": "dependencies",
        "depName": "@datapunt/matomo-tracker-react",
        "currentValue": "0.3.0",
        "datasource": "npm",
        "prettyDepType": "dependency",
        "lockedVersion": "0.3.0",
        "depIndex": 0,
        "updates": [
          {
            "bucket": "non-major",
            "newVersion": "0.5.1",
            "newValue": "0.5.1",
            "releaseTimestamp": "2021-08-11T07:56:06.110Z",
            "newMajor": 0,
            "newMinor": 5,
            "updateType": "minor",
            "branchName": "renovate-datapunt-matomo-tracker-react-0.x"
          }
        ],
        "warnings": [],
        "versioning": "npm",
        "sourceUrl": "https://github.com/Amsterdam/matomo-tracker",
        "sourceDirectory": "packages/react",
        "currentVersion": "0.3.0",
        "isSingleVersion": true,
        "fixedVersion": "0.3.0"
      }
    ],
    "prettyNewVersion": "v0.5.1",
    "displayFrom": "0.3.0",
    "displayTo": "0.5.1",
    "displayPending": "",
    "isRange": false
  }
}
```

</p></details>

**Reply by rarkins** [reply]
Unfortunately this seems very hard to troubleshoot or even understand for anyone not in your environment. I'm guessing the "wrong" name comes via a direct API call without filters. We access APIs directly, not via git, but to reverse engineer Git's filters would also be undesirable.

**Reply by mikaello** [reply]
I see. Anyway, thanks for all your tips and suggestions, and that you took your time to try to help me figure out a solution.

I agree with you that this seems like a difficult case, and that we probably have to find another way to solve this issue without making Renovate work with Git filters.

**Comment by rarkins** [comment]
Converted from issue to discussion
